README
The Python Imaging Library
$Id$

Release 1.2a0 (January 8, 2011)

====================================================================
The Python Imaging Library 1.2a0
====================================================================

Contents
--------

+ Introduction
+ Support
+ Software License
+ Build instructions (all platforms)
  - Additional notes for Mac OS X
  - Additional notes for Windows

--------------------------------------------------------------------
Introduction
--------------------------------------------------------------------

The Python Imaging Library (PIL) adds image processing capabilities
to your Python environment.  This library provides extensive file
format support, an efficient internal representation, and powerful
image processing capabilities.

This source kit has been built and tested with Python 2.3 and newer,
on Windows, Mac OS X, and major Unix platforms.  The kit installs
under Python 3.1, but doesn't provide much functionality yet.

The main distribution site for this software is:

        http://www.pythonware.com/products/pil/

That site also contains information about free and commercial support
options, PIL add-ons, answers to frequently asked questions, and more.


Development versions (alphas, betas) are available here:

        http://effbot.org/downloads/


The full PIL handbook is not included in this distribution; to get the
latest version, check:

        http://www.pythonware.com/library/
        http://effbot.org/books/imagingbook/ (drafts)

However, some reference material is available in the Docs directory.


For installation and licensing details, see below.


--------------------------------------------------------------------
Support
--------------------------------------------------------------------

For support and general questions on the Python Imaging Library, send
e-mail to the Image SIG mailing list:

        image-sig@python.org

You can join the Image SIG by sending a mail to:

        image-sig-request@python.org

Put "subscribe" in the message body to automatically subscribe to the
list, or "help" to get additional information.  Alternatively, you can
send your questions to the Python mailing list, python-list@python.org,
or post them to the newsgroup comp.lang.python.  DO NOT SEND SUPPORT
QUESTIONS TO PYTHONWARE ADDRESSES.


--------------------------------------------------------------------
Software License
--------------------------------------------------------------------

The Python Imaging Library is

Copyright (c) 1997-2011 by Secret Labs AB
Copyright (c) 1995-2011 by Fredrik Lundh

By obtaining, using, and/or copying this software and/or its
associated documentation, you agree that you have read, understood,
and will comply with the following terms and conditions:

Permission to use, copy, modify, and distribute this software and its
associated documentation for any purpose and without fee is hereby
granted, provided that the above copyright notice appears in all
copies, and that both that copyright notice and this permission notice
appear in supporting documentation, and that the name of Secret Labs
AB or the author not be used in advertising or publicity pertaining to
distribution of the software without specific, written prior
permission.

SECRET LABS AB AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO
THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS.  IN NO EVENT SHALL SECRET LABS AB OR THE AUTHOR BE LIABLE FOR
ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


--------------------------------------------------------------------
Build instructions (all platforms)
--------------------------------------------------------------------

For a list of changes in this release, see the CHANGES document.

0. If you're in a hurry, try this:

        $ tar xvfz Imaging-1.2a0.tar.gz
        $ cd Imaging-1.2a0
        $ python setup.py install

   If you prefer to know what you're doing, read on.


1. Prerequisites.

   If you need any of the features described below, make sure you
   have the necessary libraries before building PIL.

   feature              library
   -----------------------------------------------------------------
   JPEG support         libjpeg (6b or 8b)

                        http://www.ijg.org
                        http://www.ijg.org/files/jpegsrc.v6b.tar.gz
                        http://www.ijg.org/files/jpegsrc.v8b.tar.gz
                        ftp://ftp.uu.net/graphics/jpeg/

   PNG support          zlib (1.2.3 or later is recommended)

                        http://www.gzip.org/zlib/

   OpenType/TrueType    freetype2 (2.3.9 or later is recommended)
   support              
                        http://www.freetype.org
                        http://freetype.sourceforge.net  

   CMS support          littleCMS (1.1.5 or later is recommended)
   support              
                        http://www.littlecms.com/

   If you have a recent Linux version, the libraries provided with the
   operating system usually work just fine.  If some library is
   missing, installing a prebuilt version (jpeg-devel, zlib-devel,
   etc) is usually easier than building from source.  For example, for
   Ubuntu 9.10 (karmic), you can install the following libraries:

       sudo apt-get install libjpeg62-dev
       sudo apt-get install zlib1g-dev
       sudo apt-get install libfreetype6-dev
       sudo apt-get install liblcms1-dev

   For Tkinter development, you also need:

       sudo apt-get install python-tk
       sudo apt-get install tk-dev

   If you're using Mac OS X, you can use the 'fink' tool to install
   missing libraries (also see the Mac OS X section below).

   Similar tools are available for many other platforms.
   

2. If you didn't build Python from sources, make sure you have
   Python's build support files on your machine.  If you've down-
   loaded a prebuilt package (e.g. a Linux RPM), you probably
   need additional developer packages.  Look for packages named
   "python-dev", "python-devel", or similar.  For example, for
   Ubuntu 9.10 (karmic), use the following command:

       sudo apt-get install python-dev


3. When you have everything you need, unpack the PIL distribution
   (the file Imaging-1.2a0.tar.gz) in a suitable work directory:

        $ cd MyExtensions # example
        $ gunzip Imaging-1.2a0.tar.gz
        $ tar xvf Imaging-1.2a0.tar


4. Build the library.  We recommend that you do an in-place build,
   and run the self test before installing.

        $ cd Imaging-1.2a0
        $ python setup.py build_ext -i
        $ python selftest.py

   During the build process, the setup.py script will display a
   summary report that lists what external components it found.
   The self test will display a similar report, with what external
   components the tests found in the actual build files:

        ----------------------------------------------------------------
        PIL 1.2a0 SETUP SUMMARY
        ----------------------------------------------------------------
        *** TKINTER support not available (Tcl/Tk 8.5 libraries needed)
        --- JPEG support available
        --- ZLIB (PNG/ZIP) support available
        --- FREETYPE support available
        ----------------------------------------------------------------

   Make sure that the optional components you need are included.

   If the build script won't find a given component, you can edit the
   setup.py file and set the appropriate ROOT variable.  For details,
   see instructions in the file.

   If the build script finds the component, but the tests cannot
   identify it, try rebuilding *all* modules:

        $ python setup.py clean
        $ python setup.py build_ext -i


5. If the setup.py and selftest.py commands finish without any
   errors, you're ready to install the library:

        $ python setup.py install

   (depending on how Python has been installed on your machine,
   you might have to log in as a superuser to run the 'install'
   command, or use the 'sudo' command to run 'install'.)


--------------------------------------------------------------------
Additional notes for Mac OS X
--------------------------------------------------------------------

On Mac OS X you will usually install additional software such as
libjpeg or freetype with the "fink" tool, and then it ends up in
"/sw".  If you have installed the libraries elsewhere, you may have
to tweak the "setup.py" file before building.


--------------------------------------------------------------------
Additional notes for Windows
--------------------------------------------------------------------

On Windows, you need to tweak the ROOT settings in the "setup.py"
file, to make it find the external libraries.  See comments in the
file for details.

Make sure to build PIL and the external libraries with the same
runtime linking options as was used for the Python interpreter
(usually /MD, under Visual Studio).


Note that most Python distributions for Windows include libraries
compiled for Microsoft Visual Studio.  You can get the free Express
edition of Visual Studio from:

    http://www.microsoft.com/Express/

To build extensions using other tool chains, see the "Using
non-Microsoft compilers on Windows" section in the distutils handbook:

    http://www.python.org/doc/current/inst/non-ms-compilers.html

For additional information on how to build extensions using the
popular MinGW compiler, see:

    http://mingw.org (compiler)
    http://sebsauvage.net/python/mingw.html (build instructions)
    http://sourceforge.net/projects/gnuwin32 (prebuilt libraries)



CHANGES
-*- coding: utf-8 -*-

The Python Imaging Library
$Id$

ACKNOWLEDGEMENTS: PIL wouldn't be what it is without the help of:
David Ascher, Phil Austin, Douglas Bagnall, Larry Bates, Anthony
Baxter, William Baxter, Denis Benoit, Jan Blom, Duncan Booth, Alexey
Borzenkov, Jeff Breidenbach, Roger Burnham, Zac Burns, Gene Cash,
Kevin Cazabon, Fred Clare, Greg Coats, Chris Cogdon, Greg Couch, Bill
Crutchfield, Abel Deuring, Tim Docker, Fred Drake, Graham Dumpleton,
Matthew Ellis, Eric Etheridge, Daniel Fetchinson, Robin Friedrich,
Pier Paolo Glave, Federico Di Gregorio, Markus Gritsch, Daniel
Haertle, Greg Hamilton, Mark Hammond, Bernhard Herzog, Rob Hooft, Bob
Ippolito, Jack Jansen, Bill Janssen, Edward Jones, Richard Jones,
Håkan Karlsson, Robert Kern, David Kirtley, Bob Klimek, Matthias
Klose, Andrew Kuchling, Magnus Källström, Victor Lacina, Ben Last,
Hamish Lawson, Cesare Leonardi, Andrew MacIntyre, Jan Matejek, Naveen
Michaud-Agrawal, Gordon McMillan, Skip Montanaro, Fredrik Nehr,
Russell Nelson, Luciano Nocera, Travis Oliphant, Piet van Oostrum,
Richard Oudkerk, Paul Pharr, Andres Polit, Conrado Porto Lopes Gouvêa,
Eric Raymond, Victor Reijs, Bertil Reinhammar, Nicholas Riley, Don
Rozenberg, Toby Sargeant, Barry Scott, Les Schaffer, Joel Shprentz,
Klamer Shutte, Gene Skonicki, Niki Spahiev, D. Alan Stewart, Perry
Stoll, Paul Svensson, Ulrik Svensson, Miki Tebeka, Michael van
Tellingen, Ivan Tkatchev, Dan Torop, Adam Twardoch, Rune Uhlin, Dmitry
Vasiliev, Sasha Voynow, Charles Waldman, Collin Winter, Dan Wolfe,
Ka-Ping Yee, and many others (if your name should be on this list, let
me know.)

*** Changes from release 1.1.7 to 1.2 ***

This section may not be fully complete.  For changes since this file
was last updated, see the repository revision history:

  http://bitbucket.org/effbot/pil-2009-raclette/changesets/

+ Added support for identifying WebP files.

+ Don't choke on Unicode strings when using bitmap fonts; render
  replace characters instead.

+ Added Oliver Tonnhofer's PNG and Quantization work:

  - Added support for PNG8 images with transparency
  - Added a new color quantizer: Image.FASTOCTTREE
  - Added new compress_level and compress_type options to
    the PNG writer to get more control over ZIP compression

+ Add lib64 directories on Linux (from Laurence Rowe).

+ On Darwin look in /usr/x11 for freetype (from Laurence Rowe)..

+ Restored support for Python 2.2, for now.

(1.2a0-20110108 snapshot)

+ Setup now migrates all Python files if installing under Python
  3.1 or later.

+ Modules must now be imported using the full package name.  The
  PIL.pth helper has been removed.

+ Added support for identifying OpenEXR files.

+ Removed obsolete ARG codec.

+ Removed deprecated Image.offset and ImageDraw.setink/setfill
  methods.

+ All classes now inherit from object.

+ Added support for binary file streams (currently used by PNG and JPEG
  only).

+ Dropped support for versions older than 2.3, and removed many
  obsolete constructs from the code.

+ Added support for reading 16-bit RGB TIFF files.

+ Fixed numpy conversion for YCbCr images (from David Coles).

*** Changes from release 1.1.6 to 1.1.7 ***

(1.1.7 final)

+ Set GIF loop info property to the number of iterations if a NETSCAPE
  loop extension is present, instead of always setting it to 1 (from
  Valentino Volonghi).

(1.1.7c1 released)

+ Improved PNG compression (from Alexey Borzenkov).

+ Read interlaced PNG files (from Conrado Porto Lopes Gouvêa)

+ Added various TGA improvements from Alexey Borzenkov, including
  support for specifying image orientation.

+ Bumped block threshold to 16 megabytes, made size estimation a bit
  more accurate.  This speeds up allocation of large images.

+ Fixed rounding error in ImagingDrawWideLine.

  "gormish" writes: ImagingDrawWideLine() in Draw.c has a bug in every
  version I've seen, which leads to different width lines depending on
  the order of the points in the line. This is especially bad at some
  angles where a 'width=2' line can completely disappear.

+ Added support for RGBA mode to the SGI module (based on code by
  Karsten Hiddemann).

+ Handle repeated IPTC tags (adapted from a patch by Eric Bruning).

  Eric writes: According to the specification, some IPTC tags can be
  repeated, e.g., tag 2:25 (keywords). PIL 1.1.6 only retained the last
  instance of that tag. Below is a patch to store all tags. If there are
  multiple tag instances, they are stored in a (python) list. Single tag
  instances remain as strings.

+ Fixed potential crash in ImageFilter for small target images
  (reported by Zac Burns and Daniel Fetchinson).

+ Use BMP instead of JPEG as temporary show format on Mac OS X.

+ Fixed putpixel/new for I;16 with colors > 255.

+ Added integer power support to ImagingMath.

+ Added limited support for I;16L mode (explicit little endian).

+ Moved WMF support into Image.core; enable WMF rendering by default
  if renderer is available.

+ Mark the ARG plugin as obsolete.

+ Added version query mechanism to ImageCms and ImageFont, for
  debugging.

+ Added (experimental) ImageCms function for fetching the ICC profile
  for the current display (currently Windows only).

  Added HWND/HDC support to ImageCms.get_display_profile().

+ Added WMF renderer (Windows only).

+ Added ImagePointHandler and ImageTransformHandler mixins; made
  ImageCmsTransform work with im.point.

+ Fixed potential endless loop in the XVThumbnail reader (from Nikolai
  Ugelvik).

+ Added Kevin Cazabon's pyCMS package.

  The C code has been moved to _imagingcms.c, the Python interface
  module is installed as PIL.ImageCMS.

  Added support for in-memory ICC profiles.

  Unified buildTransform and buildTransformFromOpenProfiles.

  The profile can now be either a filename, a profile object, or a
  file-like object containing an in-memory profile.

  Additional fixes from Florian Böch:

    Very nice - it just needs LCMS flags support so we can use black

PILusm.cpp
#define COPYRIGHTINFO "\
PILusm, a gaussian blur and unsharp masking library for PIL\n\
By Kevin Cazabon, copyright 2003\n\
kevin_cazabon@hotmail.com\n\
kevin@cazabon.com\n\
\n\
This library is free software; you can redistribute it and/or\n\
modify it under the terms of the GNU Lesser General Public\n\
License as published by the Free Software Foundation; either\n\
version 2.1 of the License, or (at your option) any later version.\n\
\n\
This library is distributed in the hope that it will be useful,\n\
but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU\n\
Lesser General Public License for more details.\n\
\n\
You should have received a copy of the GNU Lesser General Public\n\
License along with this library; if not, write to the Free Software\n\
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA\n\
"

/////////////////////////////////////////////////////////////////////////////
// includes
/////////////////////////////////////////////////////////////////////////////
#include "Python.h"
#include "Imaging.h"


/////////////////////////////////////////////////////////////////////////////
// version information: update this before compiling for the versions you're using
/////////////////////////////////////////////////////////////////////////////
#define PILUSMVERSION        "0.6.0"
#define PILVERSION          "1.1.3"
#define PYTHONVERSION       "2.2.0"

/////////////////////////////////////////////////////////////////////////////
// version history
/////////////////////////////////////////////////////////////////////////////
/*
0.6.0   fixed/improved float radius support (oops!)
        now that radius can be a float (properly), changed radius value to
            be an actual radius (instead of diameter).  So, you should get
            similar results from PIL_usm as from other paint programs when
            using the SAME values (no doubling of radius required any more).
            Be careful, this may "break" software if you had it set for 2x
            or 5x the radius as was recommended with earlier versions.
        made PILusm thread-friendly (release GIL before lengthly operations,
            and re-acquire it before returning to Python).  This makes a huge
            difference with multi-threaded applications on dual-processor
            or "Hyperthreading"-enabled systems (Pentium4, Xeon, etc.)

0.5.0   added support for float radius values!

0.4.0   tweaked gaussian curve calculation to be closer to consistent shape
            across a wide range of radius values

0.3.0   changed deviation calculation in gausian algorithm to be dynamic
        _gblur now adds 1 to the user-supplied radius before using it so
            that a value of "0" returns the original image instead of a
            black one.
        fixed handling of alpha channel in RGBX, RGBA images
        improved speed of gblur by reducing unnecessary checks and assignments

0.2.0   fixed L-mode image support

0.1.0   initial release
*/

/////////////////////////////////////////////////////////////////////////////
// known to-do list with current version
/////////////////////////////////////////////////////////////////////////////


/////////////////////////////////////////////////////////////////////////////
// options / configuration
/////////////////////////////////////////////////////////////////////////////


/////////////////////////////////////////////////////////////////////////////
// reference
/////////////////////////////////////////////////////////////////////////////


/////////////////////////////////////////////////////////////////////////////
// structs
/////////////////////////////////////////////////////////////////////////////
typedef struct {
    PyObject_HEAD
    Imaging image;
} ImagingObject;


/////////////////////////////////////////////////////////////////////////////
// internal functions
/////////////////////////////////////////////////////////////////////////////
UINT8
clip (double in) {
  if (in >= 255.0) return (UINT8) 255;
  if (in <= 0.0) return (UINT8) 0;
  else return (UINT8) in;
}


int 
_gblur(Imaging im, Imaging imOut, float floatRadius, int channels, int padding) {
  float *maskData = NULL;
  double mult = 0.0;
  int y = 0;
  int x = 0;
  float z = 0;
  float sum = 0.0;
  float dev = 0.0;

  float *buffer = NULL;

  int *line = NULL;
  UINT8 *line8 = NULL;

  int pix = 0;
  float *newPixel = NULL;
  int channel = 0;
  int offset = 0;
  INT32 newPixelFinals;

  int radius = 0;
  float remainder = 0.0;

  ///////////////////////////////////////////////
  // do the gaussian blur
  ///////////////////////////////////////////////
  /*
  For a symmetrical gaussian blur, instead of doing a radius*radius matrix lookup,
  you get the EXACT same results by doing a radius*1 transform, followed by a 1*radius transform.
  This reduces the number of lookups exponentially (10 lookups per pixel for a radius of 5
  instead of 25 lookups).  So, we blur the lines first, then we blur the resulting columns.
  */

  // first, round radius off to the next higher integer and hold the remainder
  // this is used so we can support float radius values properly.
  remainder = floatRadius - ((int)floatRadius);
  floatRadius = (int)(floatRadius + 0.999999);

  // next, double the radius and offset by 2.0... that way "0" returns the original image instead of a black one.
  // we multiply it by 2.0 so that it is a true "radius", not a diameter (the results match other
  // paint programs closer that way too).
  radius = (int)((floatRadius * 2.0) + 2.0);

  // create the maskData for the gaussian curve
  maskData = new float[radius];
  for (x = 0; x < radius; x++) {
    z = (float(x + 2)/((float)radius));
    dev = 0.5 + (((float)(radius * radius)) * 0.001); // you can adjust this factor to change the shape/center-weighting of the gaussian
    maskData[x] = (float) pow((1.0 / sqrt(2.0 * 3.14159265359 * dev)), ((-(z - 1.0) * -(x - 1.0)) / (2.0 * dev)));
  }

  // if there's any remainder, multiply the first/last values in MaskData it.
  // this allows us to support float radius values.
  if (remainder > 0.0) {
    maskData[0] *= remainder;
    maskData[radius - 1] *= remainder;
  }

  for (x = 0; x < radius; x++) {
    // this is done separately now due to the correction for float radius values above
    sum += maskData[x];
  }

  for (int i = 0; i < radius; i++) {
    maskData[i] *= (1.0/sum);
    //printf("%f\n", maskData[i]); // to see the maskData values, if you're interested.
  }

  // create a temporary memory buffer for the data for the first pass
  // memset the buffer to 0 so we can use it directly with +=
  buffer = new float [im->xsize * im->ysize * channels];  // don't bother about alpha/padding
  
  if (buffer == NULL) {
    return -1;
  }

  Py_BEGIN_ALLOW_THREADS  // be nice to other threads while you go off to lala land

  memset(buffer, 0, sizeof(buffer));

  // perform a blur on each line, and place in the temporary storage buffer
  newPixel = new float[4];
  for (y = 0; y < im->ysize; y++) {
    if (channels == 1 && im->image8 != NULL) {
      line8 = (UINT8*) im->image8[y];
    }
    else {
      line = im->image32[y];
    }
    for (x = 0; x < im->xsize; x++) {
      newPixel[0] = newPixel[1] = newPixel[2] = newPixel[3] = 0;
      // for each neighbor pixel, factor in its value/weighting to the current pixel
      for (pix = 0; pix < radius; pix++) {
        // figure the offset of this neighbor pixel
        offset = (int) ((-((float)radius / 2.0) + (float)pix) + 0.5);
        if (x + offset < 0) offset = -x;
        else if (x + offset >= im->xsize) offset = im->xsize - x - 1;

        // add (neighbor pixel value * maskData[pix]) to the current pixel value
        if (channels == 1) {
          buffer[(y * im->xsize) + x] += ((float) ((UINT8*)&line8[x + offset])[0]) * (maskData[pix]);
        }
        else {
          for (channel = 0; channel < channels; channel++){
            buffer[(y * im->xsize * channels) + (x * channels) + channel] += ((float) ((UINT8*)&line[x + offset])[channel]) * (maskData[pix]);
          }
        }
      }
    }
  }

  // perform a blur on each column in the buffer, and place in the output image
  for (x = 0; x < im->xsize; x++) {
    for (y = 0; y < im->ysize; y++) {
      newPixel[0] = newPixel[1] = newPixel[2] = newPixel[3] = 0;
      // for each neighbor pixel, factor in its value/weighting to the current pixel
      for (pix = 0; pix < radius; pix++) {
        // figure the offset of this neighbor pixel
        offset = (int)(-((float)radius / 2.0) + (float)pix + 0.5);
        if (y + offset < 0) offset = -y;
        else if (y + offset >= im->ysize) offset = im->ysize - y - 1;

        // add (neighbor pixel value * maskData[pix]) to the current pixel value
        for (channel = 0; channel < channels; channel++){
          newPixel[channel] += (buffer[((y + offset) * im->xsize * channels) + (x * channels) + channel]) * (maskData[pix]);
        }
      }
      // if the image is RGBX or RGBA, copy the 4th channel data to newPixel, so it gets put in imOut
      if (strcmp(im->mode, "RGBX") == 0 || strcmp(im->mode, "RGBA") == 0) {
        newPixel[3] = ((UINT8*)&line[x + offset])[3];
      }
      // pack the channels into an INT32 so we can put them back in the PIL image
      newPixelFinals = 0;
      if (channels == 1) {
        newPixelFinals = clip(newPixel[0]);
      }
      else {  // for RGB, the fourth channel isn't used anyways, so just pack a 0 in there, this saves checking the mode for each pixel.
        // this doesn't work on little-endian machines... fix it!
        newPixelFinals = clip(newPixel[0]) | clip(newPixel[1]) << 8 | clip(newPixel[2]) << 16 | clip(newPixel[3]) << 24;
      }

      // set the resulting pixel in imOut
      if (channels == 1) {
        imOut->image8[y][x] = (UINT8) newPixelFinals;
      }
      else {
        imOut->image32[y][x] = newPixelFinals;   
      }
    }
  }
  free(newPixel);

  // free the buffer
  free(buffer);

  Py_END_ALLOW_THREADS  // get the GIL back so Python knows who you are

  // return 0 on success
  return 0;
}

/////////////////////////////////////////////////////////////////////////////
// Python callable functions
/////////////////////////////////////////////////////////////////////////////
static PyObject *
versions (PyObject *self, PyObject *args) {
  return Py_BuildValue ("sss", PILUSMVERSION, PYTHONVERSION, PILVERSION);
}

static PyObject *
about (PyObject *self, PyObject *args) {
  return Py_BuildValue("s", COPYRIGHTINFO);
}

static PyObject *
copyright (PyObject *self, PyObject *args) {
  return about(self, args);
}

static PyObject *
gblur (PyObject *self, PyObject *args) {
  long idIn = 0L;
  long idOut = 0L;

  Imaging im = NULL;
  Imaging imOut = NULL;

  float radius = 0.0;
  int result = 0;

  int channels = 0;
  int padding = 0;

  if (!PyArg_ParseTuple(args, "llf", &idIn, &idOut, &radius)) {
    return Py_BuildValue("is", -1, "ERROR: Could not parse argument tuple, check documentation.");
  }

  im = (Imaging) idIn;
  imOut = (Imaging) idOut;

  if (strcmp(im->mode, "RGB") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "RGBA") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "RGBX") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "CMYK") == 0) {
    channels = 4;
    padding = 0;
  }
  else if (strcmp(im->mode, "L") == 0) {
    channels = 1;
    padding = 0;
  }
  else {
    return Py_BuildValue("is", -1, "ERROR: only RGB, RGBA, RGBX, CMYK and L mode images supported.");
  }

  result = _gblur(im, imOut, radius, channels, padding);

  if (result == 0) return Py_BuildValue("is", 0, "");
  else return Py_BuildValue("is", result, "ERROR: unknown error occurred during blur operation.");

}

static PyObject *
usm (PyObject *self, PyObject *args) {
  long idIn = 0L;
  long idOut = 0L;
  float radius = 0.0;
  int percent = 0;
  int threshold = 0;

  Imaging im = NULL;
  Imaging imOut = NULL;

  int result = 0;
  int channel = 0;
  int channels = 0;
  int padding = 0;

  int x = 0;
  int y = 0;

  int *lineIn = NULL;
  int *lineOut = NULL;
  UINT8 *lineIn8 = NULL;
  UINT8 *lineOut8 = NULL;

  int diff = 0;

  INT32 newPixel = 0;

  if (!PyArg_ParseTuple(args, "llfii", &idIn, &idOut, &radius, &percent, &threshold)) {
    return Py_BuildValue("is", -1, "ERROR: Could not parse argument tuple, check documentation.");
  }

  im = (Imaging) idIn;
  imOut = (Imaging) idOut;

  if (strcmp(im->mode, "RGB") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "RGBA") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "RGBX") == 0) {
    channels = 3;
    padding = 1;
  }
  else if (strcmp(im->mode, "CMYK") == 0) {
    channels = 4;
    padding = 0;
  }
  else if (strcmp(im->mode, "L") == 0) {
    channels = 1;
    padding = 0;
  }
  else {
    return Py_BuildValue("is", -1, "ERROR: only RGB, RGBA, RGBX, CMYK and L mode images supported.");
  }

  /////////////////////////////////////
  // first, do a gaussian blur on the image, putting results in imOut temporarily
  /////////////////////////////////////
  result = _gblur(im, imOut, radius, channels, padding);
  if (result != 0) return Py_BuildValue("is", result, "ERROR: unknown error occured during _gblur portion of usm");

  /////////////////////////////////////
  // now, go through each pixel, compare "normal" pixel to blurred pixel.
  // if the difference is more than threshold values, apply the OPPOSITE 
  // correction to the amount of blur, multiplied by percent.
  /////////////////////////////////////

  Py_BEGIN_ALLOW_THREADS  // be nice to other threads while you go off to lala land

  for (y = 0; y < im->ysize; y++) {
    if (channels == 1) {
      lineIn8 = im->image8[y];
      lineOut8 = imOut->image8[y];
    }
    else {
      lineIn = im->image32[y];
      lineOut = imOut->image32[y];
    }
    for (x = 0; x < im->xsize; x++) {
      newPixel = 0;
      // compare in/out pixels, apply sharpening
      if (channels == 1) {
        diff = ((UINT8*)&lineIn8[x])[0] - ((UINT8*)&lineOut8[x])[0];
        if (abs(diff) > threshold) {
          // add the diff*percent to the original pixel
          imOut->image8[y][x] = clip((((UINT8*)&lineIn8[x])[0]) + (diff * ((float)percent)/100.0));
        }
        else {
          // newPixel is the same as imIn
          imOut->image8[y][x] = ((UINT8*)&lineIn8[x])[0];
        }
      }

      else {
        for (channel = 0; channel < channels; channel++){
          diff = (float) ((((UINT8*)&lineIn[x])[channel]) - (((UINT8*)&lineOut[x])[channel]));
          if (abs(diff) > threshold) {
            // add the diff*percent to the original pixel
            // this may not work for little-endian systems, fix it!
            newPixel = newPixel | clip ( (float)(((UINT8*)&lineIn[x])[channel]) + (diff * (((float)percent/100.0))) ) << (channel * 8);
          }
          else {
            // newPixel is the same as imIn
            // this may not work for little-endian systems, fix it!
            newPixel = newPixel | ((UINT8*)&lineIn[x])[channel] << (channel * 8);
          }
        }
        if (strcmp(im->mode, "RGBX") == 0 || strcmp(im->mode, "RGBA") == 0) {
          // preserve the alpha channel
          // this may not work for little-endian systems, fix it!
          newPixel = newPixel | ((UINT8*)&lineIn[x])[channel] << 24;
        }
        imOut->image32[y][x] = newPixel;  
      }
    }
  }

  Py_END_ALLOW_THREADS  // get the GIL back so Python knows who you are

  return Py_BuildValue("is", 0, "");
}

/////////////////////////////////////////////////////////////////////////////
// Python interface setup
/////////////////////////////////////////////////////////////////////////////
static PyMethodDef PILusm_methods[] = {
  // pyCMS info
  {"versions", versions, 1, ".versions()"},
  {"about", about, 1, ".about()"},
  {"copyright", copyright, 1, ".copyright()"},

  {"gblur", gblur, 1, ".gblur(imIn.im.id, imOut.im.id, radius)"},
  {"usm", usm, 1, ".usm(imIn.im.id, imOut.im.id, radius, percent, threshold)"},

  {NULL, NULL}
};


extern "C" {
  void initPILusm(void)
  {
    Py_InitModule("PILusm", PILusm_methods);
  }
}
